addons/web_editor/static/tests/test_wysiwyg_collaboration.js

```diff
@@ -0,0 +1,786 @@
+                        // dataChannel is done through `openDataChannel` and we
+                        // do not want to simulate the events that thrigger the
+                        // openning of the dataChannel.
+                        if (args[0] === 'ptp_join') {
+                            return;
+                        }
+                        this._super(...args);
+                    },
+                     _getPtpClients() {
+                        return pool.peers[peerId]._connections.map((peer) => {
+                            return { id: peer.peerId }
+                        });
+                     },
+                    async _channelNotify(peerId, transportPayload) {
+                        if (!pool.peers[peerId].isOnline) return;
+                        pool.peers[peerId].wysiwyg.ptp.handleNotification(structuredClone(transportPayload));
+                    },
+
+                    _createClient() {
+                        throw new Error('Should not be called.');
+                    },
+                    _addIceCandidate() {
+                        throw new Error('Should not be called.');
+                    },
+                    _recoverConnection() {
+                        throw new Error('Should not be called.');
+                    },
+                    _killPotentialZombie() {
+                        throw new Error('Should not be called.');
+                    },
+                });
+                return ptp;
+            },
+            _getCurrentRecord() {
+                return {
+                    id: 1,
+                    fake_field: pool.lastRecordSaved,
+                }
+            },
+            _getCollaborationClientAvatarUrl() {
+                return '';
+            },
+        });
+        pool.peers[peerId] = new PeerTest({
+            peerId,
+            wysiwyg,
+            iframe,
+            wrapper: peerWysiwygWrapper,
+            pool,
+        });
+    }
+    return pool;
+}
+function removePeers(peers) {
+    for (const peer of Object.values(peers)) {
+        peer.wysiwyg.destroy();
+        peer.wrapper.remove();
+    }
+}
+
+QUnit.module('web_editor', {
+    before() {
+        patch(Wysiwyg, 'web_editor', {
+            activeCollaborationChannelNames: {
+                has: () => false,
+                add: () => {},
+                delete: () => {},
+            }
+        });
+    },
+    after() {
+        unpatch(Wysiwyg, "web_editor");
+    }
+}, () => {
+    QUnit.module('Collaboration', {}, () => {
+        /**
+         * Detect stale when <already focused | not already focused>
+         */
+        QUnit.module('Focus', {}, () => {
+            QUnit.test('Focused client should not receive step if no data channel is open', async (assert) => {
+                assert.expect(3);
+                const pool = await createPeers(['p1', 'p2', 'p3']);
+                const peers = pool.peers;
```

Comment:
I've resolved most of the comments of this PR so that we can move forward but this one is kind of trollish... You're telling me you don't see how `p1`, `p2` and `p3` are more readable than `peers.p1`,`peers.p2` and `peers.p3` when they appear on almost _all_ the lines of _all_ the tests ?